### PR TITLE
Fetch data from UMA server

### PR DESCRIPTION
**Pull request by holte**
This change will cause feature usage data to get fetched from an export server which can query the data, rather than reading from the JSON logs that we intend to stop producing.

### CODE REVIEW

**Comment by ebidel** [comment]
This is great. I'll give it a look!

How will the app have access to the endpoint? Right now I get 403 if I hit the constructed URL directly.

**Comment by holte** [comment]
Right now the request is only allowed if it comes from:

a) An admin of the uma-export cloud project
b) google.com:chromestatus app (from X-Appengine-Inbound-Appid)

File: admin.py [hunk 1]
```diff
@@ -42,183 +42,99 @@
...
-          day_percentage=(num_hits * 1.0 / total_pages)
+          day_percentage=rate
           #rolling_percentage=
           )
       entity.put()
 
+  def FetchAndSaveDate(self, date):
+    data, response_code = self._FetchData(date)
+    if response_code == 200:
+      self._SaveData(data, yesterday, histogram_id)
```
Comment (ebidel) [inline admin.py]: yesterday -> date

Comment (ebidel) [inline admin.py]: `histogram_id` looks left over?

File: admin.py [hunk 2]
```diff
@@ -42,183 +42,99 @@
...
+          date=date,
           #hits=num_hits,
           #total_pages=total_pages,
-          day_percentage=(num_hits * 1.0 / total_pages)
+          day_percentage=rate
           #rolling_percentage=
           )
       entity.put()
 
+  def FetchAndSaveDate(self, date):
```
Comment (ebidel) [inline admin.py]: `FetchAndSaveDate` -> `FetchAndSaveData`

**Comment by ebidel** [comment]
Any ideas on how I should go about testing this? I believe GAE only uses the cron for the default app. It would be nice to test this flow before switching it to prod.

**Comment by holte** [comment]
I think you could make a request directly to cron handler, and trigger it
that way, though that would still be modifying the same datastore as your
product.  Better would be if you have a separate cloud project for staging,
I could also whitelist that one.  I'll look and see if there is a way I can
get the right authorization provided from a dev_appserver

**Comment by holte** [comment]
Ok, I added some logic that will read a cookie from file and provide that on requests when running from local dev_appserver.py.  You just need to copy the cookie value from devtools after visiting https://uma-export.appspot.com/login and save that to the cookie file.  That should let you test it with dev_appserver.py

I haven't been able to get dev_appserver.py to work for me due to some package conflicts with appengine though, so I haven't tested it myself.

**Comment by ebidel** [comment]
Hey @holte, https://uma-export.appspot.com/login is 404ing for me.

**Comment by holte** [comment]
If you get a 404, then you've probably logged in.  You should still be able
to grab your cookies from DevTools > Application > cookies

**Comment by ebidel** [comment]
Got it!

File: admin.py [hunk 3]
```diff
@@ -42,183 +42,115 @@
...
-    for val in values_list:
-      bucket_range, hits_string = val.split('=') # e.g. "10-11=2175995"
+  def _FetchData(self, date):
+    params = '?end_date=%s&day_count=1' % date.strftime('%Y%m%d')
+    result = _FetchWithCookie(UMA_QUERY_SERVER + self.query_name + params)
 
-      parts = bucket_range.split('-')
+    if (result.status_code != 200):
+      logging.error('Unable to retrieve histograms data.')
+      return [None, 404]
```
Comment (ebidel) [inline admin.py]: `(None, 404)` to be consistent with below

File: admin.py [hunk 4]
```diff
@@ -42,183 +42,115 @@
...
 
-      beginning_range = int(parts[0])
-      end_range = int(parts[1])
+    json_content = result.content.split('\n', 1)[1]
```
Comment (ebidel) [inline admin.py]: Let's do:

```
    json_content = result.content.split('\n', 1)[1]
    j = json.loads(json_content)
    if not j.has_key('r'):
      logging.error(
          '%s results do not have an "r" key in the response' % self.query_name)
      return (None, 404)

    return (j['r'], result.status_code)
```

**Comment by holte** [comment]
Fixed the issue with featureobserver.cssproperties handler.

File: admin.py [hunk 5]
```diff
@@ -42,183 +42,115 @@
...
 
+  def FetchAndSaveData(self, date):
+    data, response_code = self._FetchData(date)
+    if response_code == 200:
+      self._SaveData(data, date)
+    return response_code
+
+
+UMA_QUERIES = [
+  UmaQuery(query='featureobserver',
```
Comment (holte) [inline admin.py]: Done

File: admin.py (hunk 4 above)
Comment (holte) [inline admin.py]: Done

File: admin.py (hunk 3 above)
Comment (holte) [inline admin.py]: Done

File: admin.py [hunk 6]
```diff
@@ -42,183 +42,115 @@
...
+    logging.error(e)
+  return urlfetch.fetch(url, headers={'cookie': 'SACSID=' + cookie})
 
-    # Response format is "bucket-bucket+1=hits".
-    # Example: 10-11=2175995,11-12=56635467,12-13=2432539420
-    #values_list = data['kTempHistograms'][CSSPROPERITES_BS_HISTOGRAM_ID]['b'].split(',')
-    values_list = data['kTempHistograms'][histogram_id]['b'].split(',')
 
-    #sum_total = int(data['kTempHistograms'][CSSPROPERITES_BS_HISTOGRAM_ID]['s']) # TODO: use this.
+class UmaQuery(Object):
```
Comment (holte) [inline admin.py]: Done

File: admin.py [hunk 7]
```diff
@@ -42,183 +42,115 @@
...
-      return
+def _FetchWithCookie(self, url):
+  if settings.PROD:
+    return urlfetch.fetch(url)
+  try:
+    with open(COOKIE_FILENAME, 'r') as f:
+      cookie = f.readline()
+  except IOError, e:
+    logging.error(e)
+  return urlfetch.fetch(url, headers={'cookie': 'SACSID=' + cookie})
```
Comment (holte) [inline admin.py]: Done

File: admin.py [hunk 8]
```diff
@@ -42,183 +42,115 @@
...
-    FEATURE_OBSERVER_BS_HISTOGRAM_ID: models.FeatureObserver,
-  }
 
-  def _SaveData(self, data, yesterday, histogram_id):
-    try:
-      model_class = self.MODEL_CLASS[histogram_id]
-    except Exception, e:
-      logging.error('Invalid CSS property bucket id used: %s' % histogram_id)
-      return
+def _FetchWithCookie(self, url):
```
Comment (holte) [inline admin.py]: Done

File: admin.py [hunk 9]
```diff
@@ -42,183 +42,121 @@
...
-      del properties_dict[PAGE_VISITS_BUCKET_ID]
-    else:
-      total_pages = sum(properties_dict.values())
+  def _SaveData(self, data, date):
+    property_map = self.property_map_class.get_all()
 
-    property_map = models.CssPropertyHistogram.get_all()
-    if histogram_id == FEATURE_OBSERVER_BS_HISTOGRAM_ID:
-      property_map = models.FeatureObserverHistogram.get_all()
+    for bucket_str, rate in data:
```
Comment (ebidel) [inline admin.py]: `for bucket_str, rate in data.iteritems():`

**Comment by ebidel** [comment]
Also confirmed `featureobserver.cssproperties` is pulling properly.

File: admin.py (hunk 9 above)
Comment (holte) [inline admin.py]: Done

**Comment by ebidel** [comment]
@holte 

Seeing 403s from the server:
screen shot 2016-09-20 at 11 42 30 am

**Comment by holte** [comment]
It looks like the app_id on those request are 's~cr-status' where the
export server was expecting 'google.com:chromestatus'.  Are both of those
app_id's for chromestatus or should I just make the export server only
accept the  's~cr-status'  one?

**Pull request merged**